tools/distpartitioning/globalids.py

```diff
@@ -0,0 +1,279 @@
+    is not present at the current rank, this function retrieves their shuffle_global_ids from the owning
+    rank
+
+    Parameters: 
+    -----------
+    rank : integer
+        rank of the process
+    world_size : integer
+        total no. of ranks configured
+    global_nids_ranks : list
+        list of numpy arrays (of global_nids), index of the list is the rank of the process
+                    where global_nid <-> shuffle_global_nid mapping is located. 
+    node_data : numpy ndarray, integers
+        node data with additional columns inserted
+
+    Returns:
+    --------
+    numpy ndarray
+        where the column-0 are global_nids and column-1 are shuffle_global_nids which are retrieved
+        from other processes. 
+    """
+    #build a list of sizes (lengths of lists)
+    sizes = [len(x) for x in global_nids_ranks]
+
+    #compute total_nodes whose mappings should be resolved, between orig-node-id <-> global-id
+    total_nodes = np.sum(sizes)
+    if (total_nodes == 0):
+        print('Rank: ', rank, ' -- All mappings are present locally... No need for to send any info.')
+        return None
+
+    #determine the no. of global_node_ids to send and receive and perform alltoall
+    send_counts = list(torch.Tensor(sizes).type(dtype=torch.int64).chunk(world_size))
+    recv_counts = list(torch.zeros([world_size], dtype=torch.int64).chunk(world_size))
+    alltoall_cpu(rank, world_size, recv_counts, send_counts)
+
+    #allocate buffers to receive node-ids
+    recv_nodes = []
+    for i in recv_counts:
+        recv_nodes.append(torch.zeros([i.item()], dtype=torch.int64))
+
+    #form the outgoing message
+    send_nodes = []
+    for i in range(world_size):
+        send_nodes.append(torch.Tensor(global_nids_ranks[i]).type(dtype=torch.int64))
+
+    #send-recieve messages
+    alltoallv_cpu(rank, world_size, recv_nodes, send_nodes)
+
+    #TODO: This code is not needed, and will be same as the sizes of the sent global node ids
+    # in the very first exchange. Please remove this piece of code. 
+    # for each of the received orig-node-id requests lookup and send out the global node id
+    send_sizes = [len(x.tolist()) for x in recv_nodes]
+    send_counts = list(torch.Tensor(send_sizes).type(dtype=torch.int64).chunk(world_size))
+    recv_counts = list(torch.zeros([world_size], dtype=torch.int64).chunk(world_size))
+    alltoall_cpu( rank, world_size, recv_counts, send_counts)
+
+    # allocate buffers to receive global-ids
+    recv_shuffle_global_nids = []
+    for i in recv_counts:
+        recv_shuffle_global_nids.append(torch.zeros([i.item()], dtype=torch.int64))
+
+    # Use node_data to lookup global id to send over.
+    send_nodes = []
+    for proc_i_nodes in recv_nodes:
+        #list of node-ids to lookup
+        global_nids = proc_i_nodes.numpy()
+        if (len(global_nids) != 0):
+            common, ind1, ind2 = np.intersect1d(node_data[:,3], global_nids, return_indices=True)
```

Comment:
do we need `intersect1d` here? based on our this morning's discussion, the lookup on the owner should be pretty simple.